### PR TITLE
fix(#7288): ignore short search words

### PR DESCRIPTION
**Pull request by garethbowen**
# Description

#7288 

# Code review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [ ] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# Compose URLs

If Build CI hasn't passed, these may 404:

* [Core](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:7288-short-keywords-miss-results/docker-compose/cht-core.yml)
* [CouchDB Single](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:7288-short-keywords-miss-results/docker-compose/cht-couchdb.yml)
* [CouchDB Cluster](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:7288-short-keywords-miss-results/docker-compose/cht-couchdb-clustered.yml)

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

**Comment by garethbowen** [comment]
@latin-panda Can I get a review for this please? It's split into 3 commits to make the review easier - 1 just for the fix, 1 for a refactor of the code, and 1 for sonar.

It fixes the bug and helps avoid a significant performance issue, but has a potential unintended consequence that before you could actually search for "ga" and get "gareth" - now that short search will be ignored and you will get all contacts/reports. I'm assuming that most users will be ok typing in at least 3 characters, but there's nothing in the UI to explain what's going on.

Anyway, please have a look!

File: shared-libs/search/src/generate-search-requests.js [hunk 1]
```diff
@@ -1,18 +1,14 @@
...
+
 const END_OF_ALPHABET = '\ufff0';
+const MINIMUM_SEARCH_TERM_LENGTH = 3;
 
-const getKeysArray = function(keys) {
-  return keys.map(function(t) {
-    return [ t ];
-  });
-};
+const getKeysArray = (keys) => keys.map(t => [ t ]);
```
Comment (latin-panda) [inline shared-libs/search/src/generate-search-requests.js:7]:
```suggestion
const getKeysArray = (keys) => keys.map(term => [ term ]);
```
I wonder why Sonar didn't catch single-letters variable names. Maybe the [rule is off](https://rules.sonarsource.com/javascript/RSPEC-117/?search=name)?

File: shared-libs/search/src/generate-search-requests.js [hunk 2]
```diff
@@ -65,63 +61,60 @@ const reportedDateRequest = function(filters) {
...
+  } else {
+    // use starts with
+    if (word.length < MINIMUM_SEARCH_TERM_LENGTH) {
+      return;
+    }
+    params.startkey = [ word ];
+    params.endkey = [ word + END_OF_ALPHABET ];
+  }
+  return params;
+};
```
Comment (latin-panda) [inline shared-libs/search/src/generate-search-requests.js:99]:
```suggestion
const freetextRequestParams = (word) => {
  const params = {};
  if (word.indexOf(':') !== -1) {
    // use exact match
    params.key = [ word ];
    return params;
  }
  
  // use starts with
  if (word.length < MINIMUM_SEARCH_TERM_LENGTH) {
    return;
  }
  
  params.startkey = [ word ];
  params.endkey = [ word + END_OF_ALPHABET ];
  return params;
};
```
Flatter code.

File: shared-libs/search/src/generate-search-requests.js [hunk 3]
```diff
@@ -65,63 +61,60 @@ const reportedDateRequest = function(filters) {
...
+
+const freetextRequest = (filters, view) => {
   if (filters.search) {
```
Comment (latin-panda) [inline shared-libs/search/src/generate-search-requests.js:102]: Return early to make code flatter.

File: shared-libs/search/src/generate-search-requests.js [hunk 4]
```diff
@@ -65,63 +61,60 @@ const reportedDateRequest = function(filters) {
...
     const words = filters.search.trim().toLowerCase().split(/\s+/);
```
Comment (latin-panda) [inline shared-libs/search/src/generate-search-requests.js:103]:
```suggestion
    const words = filters.search
      .trim()
      .toLowerCase()
      .split(/\s+/);
```

**Review by latin-panda — CHANGES_REQUESTED** [review]
Very nice! Just a couple of minor suggestions

File: shared-libs/search/src/generate-search-requests.js (hunk 3 above)
Comment (garethbowen) [reply shared-libs/search/src/generate-search-requests.js]: Done.

File: shared-libs/search/src/generate-search-requests.js (hunk 1 above)
Comment (garethbowen) [reply shared-libs/search/src/generate-search-requests.js]: Replaced with `key` instead of `term`.

**Review by latin-panda — APPROVED** [review]
LGTM! Thanks!

**Comment by latin-panda** [comment]
> I'm assuming that most users will be ok typing in at least 3 characters, but there's nothing in the UI to explain what's going on.

Attaching a video about this topic: https://github.com/medic/cht-core/assets/66472237/f70b3531-f3c1-400c-96f5-adda79fdc133



@garethbowen IMO it's not too bad and we shouldn't hold this improvement. 
We can add feedback like we do in the lookup fields, that we show a `searching...` label to `Type more to search..`. Can you please open a ticket.

**Comment by garethbowen** [comment]
Improvement raised here: https://github.com/medic/cht-core/issues/8832

**Pull request merged**